### PR TITLE
Handle invalid training_step losses in DDP

### LINKED ISSUE #4524: Error when disabling an optimizer with native AMP turned on
## 🐛 Bug

When running my Lightning code with:
- fp16 native AMP 
- Multiple optimizers
- One of the optimizers disabled (in this case by returning `None` for it in `training_step`)

I'm getting the following stacktrace:

```
Traceback (most recent call last):
  File "./train_stage1.py", line 353, in <module>
    trainer.fit(model)
  File "/home/wj359634/venv/lib64/python3.6/site-packages/pytorch_lightning/trainer/trainer.py", line 440, in fit
    results = self.accelerator_backend.train()
  File "/home/wj359634/venv/lib64/python3.6/site-packages/pytorch_lightning/accelerators/gpu_accelerator.py", line 54, in train
    results = self.train_or_test()
  File "/home/wj359634/venv/lib64/python3.6/site-packages/pytorch_lightning/accelerators/accelerator.py", line 68, in train_or_test
    results = self.trainer.train()
  File "/home/wj359634/venv/lib64/python3.6/site-packages/pytorch_lightning/trainer/trainer.py", line 485, in train
    self.train_loop.run_training_epoch()
  File "/home/wj359634/venv/lib64/python3.6/site-packages/pytorch_lightning/trainer/training_loop.py", line 544, in run_training_epoch
    batch_output = self.run_training_batch(batch, batch_idx, dataloader_idx)
  File "/home/wj359634/venv/lib64/python3.6/site-packages/pytorch_lightning/trainer/training_loop.py", line 713, in run_training_batch
    self.optimizer_step(optimizer, opt_idx, batch_idx, train_step_and_backward_closure)
  File "/home/wj359634/venv/lib64/python3.6/site-packages/pytorch_lightning/trainer/training_loop.py", line 453, in optimizer_step
    optimizer, batch_idx, opt_idx, train_step_and_backward_closure
  File "/home/wj359634/venv/lib64/python3.6/site-packages/pytorch_lightning/accelerators/accelerator.py", line 122, in optimizer_step
    using_lbfgs=is_lbfgs
  File "/home/wj359634/venv/lib64/python3.6/site-packages/pytorch_lightning/core/lightning.py", line 1209, in optimizer_step
    self.trainer.scaler.step(optimizer)
  File "/home/wj359634/venv/lib64/python3.6/site-packages/torch/cuda/amp/grad_scaler.py", line 318, in step
    assert len(optimizer_state["found_inf_per_device"]) > 0, "No inf checks were recorded for this optimizer."
AssertionError: No inf checks were recorded for this optimizer.
Exception ignored in: <object repr() failed>
Traceback (most recent call last):
  File "/home/wj359634/venv/lib64/python3.6/site-packages/tqdm/std.py", line 1086, in __del__
  File "/home/wj359634/venv/lib64/python3.6/site-packages/tqdm/std.py", line 1293, in close
  File "/home/wj359634/venv/lib64/python3.6/site-packages/tqdm/std.py", line 1471, in display
  File "/home/wj359634/venv/lib64/python3.6/site-packages/tqdm/std.py", line 1089, in __repr__
  File "/home/wj359634/venv/lib64/python3.6/site-packages/tqdm/std.py", line 1433, in format_dict
TypeError: 'NoneType' object is not iterable
```

### To Reproduce

(I'm hoping those are all conditions that have to be met)
Run a Lightning model with
- fp16 native AMP 
- Multiple optimizers
- One of the optimizers disabled (in this case by returning `None` for it in `training_step`)

### Expected behavior

The code should skip this optimizer

### Environment

```
* CUDA:
        - GPU:
                - Tesla V100-PCIE-32GB
        - available:         True
        - version:           10.2
* Packages:
        - numpy:             1.18.4
        - pyTorch_debug:     True
        - pyTorch_version:   1.7.0
        - pytorch-lightning: 1.0.4
        - tqdm:              4.46.0
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - ELF
        - processor:         x86_64
        - python:            3.6.8
        - version:           #1 SMP Tue Aug 25 17:23:54 UTC 2020
```

### LINKED ISSUE #4956: how to properly skip samples that cause inf/nan gradients/loss
#### tl;dr
does the approach in the code snippet below look ok, or is there a better alternative for automatically skipping few "bad" samples in the data that cause inf/nan gradients/loss? (is it a good practice altogether?)

#### details
sometimes, there is a small percentage (but annoyingly large in absolute value) of "dirty" samples in the data that cause the loss to be nan, although the neural-network architecture itself is fine and stable in terms of numerical stability.  
one approach is to automatically stop training (use `terminate_on_nan`) and then somehow isolate **all** these samples and remove them from the data permanently. but..  
sometimes we simply want to automatically skip these samples as if they never existed (perhaps with a warning), and continue training.  
I couldn't find any documentation about how to do that, nor anyone who asked this question. so i decided to ask and offer a solution I found, for others that might need it as well.  
in the end, i came up with the following approach - override `on_after_backwards` method in my lightning-module with the following code:  

#### code
```# catch inf/nan gradients and zero them, to avoid nan loss that will ruin the entire training due to possibly few bad samples
    def on_after_backward(self) -> None:
        valid_gradients = True
        for name, param in self.named_parameters():
            if param.grad is not None:
                valid_gradients = not (torch.isnan(param.grad).any() or torch.isinf(param.grad).any())
                if not valid_gradients:
                    break

        if not valid_gradients:
            log.warning(f'detected inf or nan values in gradients. not updating model parameters')
            self.zero_grad()
```

#### pros
* this code successfully identifies nan/inf gradients, and skips parameter update by zeroing gradients for the specific batch
* support multi-gpu (at least ddp which I tested). when done this way, detecting inf/nan gradients (instead of inf/nan loss), we avoid a potential cases of losing synchronization between different processes, because typically one of the processes would generate an inf loss, while the others won't. if we stop only one process from doing a backwards pass, we lose synchronization, and would stumble into a never-ending processes that wait for nothing. training stalls. when checking gradients, it is after **all** gradients in all processes have been affected by the bad inf loss. so we have synchronization.
#### cons
* can't catch bad samples that way.. need to work harder..
* might not be future proof
* clutters lightning module code (it is essentially architecture agnostic, boiler-plate code)
* perhaps there is a better way..

#### final question 
is it worth having such functionality integrated into lightning as a simple command-line-switch/parameter?

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

Fixes #4956, #5243 (DDP)
Fixes #4524 (AMP)

This PR allows choosing how to skip a step when training_step returns `NaN`, `None`, or `inf` in DDP.
We introduce a LightningModule attribute to choose the strategy (via a str/Enum):

> This requires torch>=1.7 to work. We chose not to support earlier versions because the amount of work necessary to them is probably not worth, given the number of users interested in this feature.

```console
Strategy 1: "skip_if_any"
process1 -> invalid loss (None or NaN/inf tensor)
process2 -> tensor(0.7)

It will skip the update for all processes (default). 
```

```console
Strategy 2: "never_skip"
process1 -> tensor(...) -> grad_1.1
process2 -> tensor(...) -> grad_2.1

process1 -> invalid loss
process2 -> tensor(...) -> grad_2.2

Decision: Backward without sync on p2 and all_reduce grad after: (grad_1.1 + grad_2.1 + grad_2.2) / 3
```

## TODO:

- [x] Weighted Reduction with accumulate_grad_batches  = 1
- [x] Weighted Reduction with accumulate_grad_batches > 1
- [x] Test with ddp_spawn
- [x] Test with ddp_cpu, ddp_spawn.
- [ ] Test with ddp_sharded
- [ ] Resolve Logging being NaN in progress_bar. `self.log(filter_infinite=True)`
- [x] Support AMP
- [ ] Support APEX?
- [ ] Test with Horovod (doesn't support `_register_comm_hook` ) 
- [ ] Support `+-Inf` (cant use torch.isfinite`, requires torch 1.6+)
- [ ] Add parity test to make sure gradient reduction is done properly (accumulate_grad_batches 1, 2)
- [ ] Separate test for both strategies
- [ ] Train real model to check training works works fine.
- [ ] Check behaviour in `manual_optimization` setting.
- [ ] Move to 1.2-dev as it is more a new feature.
- [ ] Test misconfigs

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes [if needed]?
- [ ] Did you write any new necessary tests [no need for typos, docs]? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified
 - [ ] **Check that target branch and milestone are aligned!**

### CODE REVIEW

File: pytorch_lightning/core/callbacks = ControlFlow(callbacks=callba [hunk 1]
```diff
@@ -0,0 +1,2 @@
+Trainer(callbacks=[callbacks])
```
Comment (carmocca) [inline pytorch_lightning/core/callbacks = ControlFlow(callbacks=callba:1]: you accidentally committed this :D

File: pytorch_lightning/core/lightning.py [hunk 2]
```diff
@@ -162,6 +164,26 @@ def automatic_optimization(self) -> bool:
         """
         return True
 
+    @property
+    def decision_on_invalid_result(self) -> Union[str, int]:
```
Comment (carmocca) [inline pytorch_lightning/core/lightning.py:168]: new users do not know about the internal result object. call it `invalid_loss_strategy`?

Comment (carmocca) [reply pytorch_lightning/core/lightning.py:168]: also return should be just `str`

Why did you not end up using an Enum?

`class InvalidLossStrategy(str, Enum)`

File: pytorch_lightning/core/lightning.py [hunk 3]
```diff
@@ -162,6 +164,26 @@ def automatic_optimization(self) -> bool:
...
+        """
+        This property is used to decide behaviour on ``training_step`` result object 
+        when using LightningDistributedDataParallel.
+
+        Can either be "skip_on_at_least_one" or "never_skip".
+        
+        "skip_on_at_least_one":
```
Comment (carmocca) [inline pytorch_lightning/core/lightning.py:175]: `skip_if_any` is shorter

File: pytorch_lightning/core/lightning.py [hunk 4]
```diff
@@ -1146,8 +1168,41 @@ def backward(self, loss, optimizer, optimizer_idx):
...
+            else:
+                loss.backward(*args, **kwargs)
+
+    def _backward_with_possible_nan_loss(self, loss: Tensor, *args, **kwargs) -> None:
+        is_loss_nan = torch.isnan(loss).int()
+        self.trainer.accelerator_backend.sync_tensor(is_loss_nan)
+        no_nan_losses = is_loss_nan == 0
+        self.trainer.model.require_backward_grad_sync = False
+
+        if not torch.isnan(loss):
```
Comment (carmocca) [inline pytorch_lightning/core/lightning.py:1189]: i think i'd move this condition to the `backward` fn above

File: pytorch_lightning/trainer/properties.py [hunk 5]
```diff
@@ -297,6 +297,11 @@ def distributed_sampler_kwargs(self):
 
         return kwargs
 
+    def print_colored_rank(self, msg):
```
Comment (carmocca) [inline pytorch_lightning/trainer/properties.py:300]: is this for debugging the draft or do you want to merge it?

if so, i'd do it separately

File: pytorch_lightning/core/lightning.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/core/lightning.py:168]: It is a LightningEnum. It should work with both str and Enum. I used the enum, so people can go to definition if they look into the code.

Comment (tchaton) [reply pytorch_lightning/core/lightning.py:168]: Renamed `invalid_loss_strategy`.

File: pytorch_lightning/core/lightning.py (hunk 3 above)
Comment (tchaton) [reply pytorch_lightning/core/lightning.py:175]: Done !

File: pytorch_lightning/trainer/properties.py (hunk 5 above)
Comment (tchaton) [reply pytorch_lightning/trainer/properties.py:300]: Yes, I will remove it when when the PR is finished and submit a better version in another PR.

File: pytorch_lightning/plugins/ddp_plugin.py [hunk 6]
```diff
@@ -68,6 +68,9 @@ def configure_ddp(self, model, device_ids):
             device_ids=device_ids,
             **self._ddp_kwargs,
         )
+        if model.module.decision_on_invalid_result == DecisionOnInvalidResult.NEVER_SKIP:
```
Comment (carmocca) [inline pytorch_lightning/plugins/ddp_plugin.py:71]: why do we need to set these here if it is already done in `_backward_with_possible_nan_loss`?

File: pytorch_lightning/trainer/training_loop.py [hunk 7]
```diff
@@ -856,6 +862,39 @@ def backward(self, result, optimizer, opt_idx, *args, **kwargs):
...
+        is_result_none = result is None
+        if isinstance(self.trainer.model, LightningDistributedDataParallel):
+            model_ref = self.trainer.get_model()
+            skip_on_at_leat_one = model_ref.decision_on_invalid_result == DecisionOnInvalidResult.SKIP_ON_AT_lEAST_ONE
+            never_skip = model_ref.decision_on_invalid_result == DecisionOnInvalidResult.NEVER_SKIP
+
+            if never_skip and is_result_none:
+                raise MisconfigurationException(
+                    "decision_on_invalid_result ``never_skip`` doesn't support returning None for training_step. "
+                    "Hint: Either switch to ``skip_on_at_least_one`` or return a Nan or Inf loss directly"
```
Comment (carmocca) [inline pytorch_lightning/trainer/training_loop.py:875]: I would do this for the user instead of raising an Exception

File: pytorch_lightning/plugins/ddp_plugin.py (hunk 6 above)
Comment (tchaton) [reply pytorch_lightning/plugins/ddp_plugin.py]: If I don't, it hangs in DDP.

Comment (tchaton) [reply pytorch_lightning/plugins/ddp_plugin.py]: Definitely need more investigation 👍

File: pytorch_lightning/trainer/training_loop.py (hunk 7 above)
Comment (tchaton) [reply pytorch_lightning/trainer/training_loop.py]: I would prefer to leave a MisConfigurationException here.

This would be triggered only when people change the property `invalid_loss_strategy` to return `never_skip`. I think they should be aware of what is happening and change their code accordingly.
I don't think Lightning should choose between the 2 modes.

File: pytorch_lightning/plugins/ddp_plugin.py (hunk 6 above)
Comment (tchaton) [reply pytorch_lightning/plugins/ddp_plugin.py]: Seemed to work when removing then.

File: pytorch_lightning/core/lightning.py [hunk 8]
```diff
@@ -1146,8 +1174,77 @@ def backward(self, loss, optimizer, optimizer_idx):
...
-        if self.trainer.train_loop.automatic_optimization or self._running_manual_backward:
-            loss.backward(*args, **kwargs)
+        loss.backward(*args, **kwargs)
+
+    def _backward_wrapper(self, loss: Tensor, optimizer: Optimizer, optimizer_idx: int, *args, **kwargs) -> None:
+        automatic_optimization = self.trainer.train_loop.automatic_optimization
+
+        if automatic_optimization or self._running_manual_backward:
+
+            is_ddp = isinstance(self.trainer.model, LightningDistributedDataParallel)
```
Comment (SeanNaren) [inline pytorch_lightning/core/lightning.py:1184]: Nice, we should leave this responsibility with the accelerator imo. such that we can do `self.backward_with_possible_nan_loss(...)` within the accelerator. Is that a possibility?

File: pytorch_lightning/core/lightning.py [hunk 9]
```diff
@@ -1146,8 +1174,77 @@ def backward(self, loss, optimizer, optimizer_idx):
...
+
+    def __create_zeros_gradients(self):
+        """
+        This function will generate zeros gradients when None for the first reduction
+        """
+        for p in self.parameters():
+            if p.requires_grad and p.grad is None:
+                p.grad = torch.zeros_like(p, device=self.device, dtype=torch.float)
+
+    def __synchronize_gradients(self, optimizer):
```
Comment (SeanNaren) [inline pytorch_lightning/core/lightning.py:1223]: Why is this needed? It's an additional overhead that is expensive to do, since I think DDP already does this internally. Any way we can remove this requirement?

**Comment by SeanNaren** [comment]
What a super neat feature @tchaton, if this was out earlier last year would've moved my entire codebase to PL when using native torch CTC :D 

Overall it would be nice for the DDP code to be encapsulated within the accelerator, maybe exposing some function to handle the syncs in DDP.

File: pytorch_lightning/core/lightning.py (hunk 8 above)
Comment (tchaton) [reply pytorch_lightning/core/lightning.py:1184]: Definitely. I will move this.

File: pytorch_lightning/trainer/training_loop.py (hunk 7 above)
Comment (carmocca) [reply pytorch_lightning/trainer/training_loop.py]: I'm talking about `return a Nan or Inf loss directly`

File: pytorch_lightning/core/lightning.py [hunk 10]
```diff
@@ -111,6 +113,8 @@ def __init__(self, *args, **kwargs):
         self._running_manual_backward = False
         self._current_hook_fx_name = None
         self._current_dataloader_idx = None
+        self._invalid_loss_strategy = InvalidLossStrategy.SKIP_IF_ANY
```
Comment (carmocca) [inline pytorch_lightning/core/lightning.py:116]: Why did you change it to be a protected attribute (instead of just the property)

File: pytorch_lightning/core/lightning.py [hunk 11]
```diff
@@ -162,6 +166,30 @@ def automatic_optimization(self) -> bool:
...
+            If result is None on at least on, it will skip
+            ``backward`` and ``optimizer_step`` for all processes
+
+        "never_skip":
+            If result loss is NaN or Inf on 1 or more processes,
+            those processes will skip backward and gradient will be
+            synchronized before running ``optimizer.step``
+
+        """
+        return self._invalid_loss_strategy
```
Comment (carmocca) [inline pytorch_lightning/core/lightning.py:187]:
```suggestion
    def invalid_loss_strategy(self) -> Union[str, InvalidLossStrategy]:
        """
        This property is used to decide the optimization behaviour when
        ``training_step`` returns an invalid loss (None, NaN, or +-Inf)
        when using DDP.

        Can either be ``"skip_if_any"`` or ``"never_skip"``.
        
        "skip_if_any":
            If one process returns an invalid loss, ``backward`` and
            ``optimizer_step`` will be skipped on all processes

        "never_skip":
            If the loss is invalid on 1 or more processes,
            those processes will skip ``backward`` and gradients will be
            synchronized before running ``optimizer.step``
        """
```

File: pytorch_lightning/core/lightning.py [hunk 12]
```diff
@@ -162,6 +166,30 @@ def automatic_optimization(self) -> bool:
...
+
+    @invalid_loss_strategy.setter
+    def invalid_loss_strategy(self, invalid_loss_strategy):
+        self._invalid_loss_strategy = invalid_loss_strategy
```
Comment (carmocca) [inline pytorch_lightning/core/lightning.py:191]: I like using a setter. We can run the input validation here:

```suggestion
    def invalid_loss_strategy(self, invalid_loss_strategy):
        allowed = list(InvalidLossStrategy)
        if invalid_loss_strategy not in allowed:
            raise MisconfigurationException(
                f'LightningModule `invalid_loss_strategy` property should be within {allowed}. Found {invalid_loss_strategy}'
            )
```

File: pytorch_lightning/core/lightning.py [hunk 13]
```diff
@@ -1146,8 +1174,77 @@ def backward(self, loss, optimizer, optimizer_idx):
...
+
+    def _backward_wrapper(self, loss: Tensor, optimizer: Optimizer, optimizer_idx: int, *args, **kwargs) -> None:
+        automatic_optimization = self.trainer.train_loop.automatic_optimization
+
+        if automatic_optimization or self._running_manual_backward:
+
+            is_ddp = isinstance(self.trainer.model, LightningDistributedDataParallel)
+            never_skip = self.invalid_loss_strategy == InvalidLossStrategy.NEVER_SKIP
+
+            if is_ddp and never_skip and automatic_optimization:
```
Comment (carmocca) [inline pytorch_lightning/core/lightning.py:1187]: call `check_invalid_loss` here instead of in `backward_with_possible_nan_loss`?

then we can rename it to `backward_with_nan_loss`

File: pytorch_lightning/core/lightning.py [hunk 14]
```diff
@@ -1146,8 +1174,77 @@ def backward(self, loss, optimizer, optimizer_idx):
...
+            self.__create_zeros_gradients()
+
+        if not self.trainer.train_loop.should_accumulate():
+            self.__synchronize_gradients(optimizer)
+
+    def __check_invalid_loss(self, loss):
+        is_invalid = torch.isnan(loss).int()
+        self.trainer.accelerator_backend.sync_tensor(is_invalid)
+        self._invalid_loss_counts.append(is_invalid)
+        return False
```
Comment (carmocca) [inline pytorch_lightning/core/lightning.py:1213]:
```suggestion
        return any(is_invalid)
```

File: pytorch_lightning/core/lightning.py [hunk 15]
```diff
@@ -1146,8 +1174,77 @@ def backward(self, loss, optimizer, optimizer_idx):
...
+
+    def __create_zeros_gradients(self):
+        """
+        This function will generate zeros gradients when None for the first reduction
+        """
+        for p in self.parameters():
+            if p.requires_grad and p.grad is None:
+                p.grad = torch.zeros_like(p, device=self.device, dtype=torch.float)
```
Comment (carmocca) [inline pytorch_lightning/core/lightning.py:1221]: any reason why you are manually setting these?

```suggestion
                p.grad = torch.zeros_like(p)
```

File: pytorch_lightning/plugins/native_amp.py [hunk 16]
```diff
@@ -36,15 +36,24 @@ def training_step(self, fx, args):
             output = fx(*args)
         return output
 
+    @staticmethod
+    def is_loss_invalid(loss):
+        return torch.isnan(loss) or torch.isfinite(loss)
```
Comment (carmocca) [inline pytorch_lightning/plugins/native_amp.py:41]: `torch.isfinite` includes nans. but we need to support 1.3 and above:

```suggestion
        if LooseVersion(torch.__version__) >= "1.6.0":
            return torch.isfinite(loss).logical_not()
        return (
            (loss != loss).any()
            or (loss == float('inf')).any() or (loss == float('-inf')).any()
        )
```

File: pytorch_lightning/plugins/sharded_plugin.py [hunk 17]
```diff
@@ -36,6 +38,13 @@ def configure_ddp(
             self, model: LightningModule, device_ids: List[int]
     ):
         self._wrap_optimizers(model)
+        if model.invalid_loss_strategy == InvalidLossStrategy.NEVER_SKIP:
+            rank_zero_warn(
+                "DDPShardedPlugin doesn't work with `never_skip` for property invalid_loss_strategy."
+                f"We will override `invalid_loss_strategy` to {InvalidLossStrategy.SKIP_IF_ANY}"
```
Comment (carmocca) [inline pytorch_lightning/plugins/sharded_plugin.py:44]:
```suggestion
                "`DDPShardedPlugin` doesn't work with `'never_skip'` as the `invalid_loss_strategy`."
                f" Overriding it to {InvalidLossStrategy.SKIP_IF_ANY.value}"
```
```

File: pytorch_lightning/trainer/configuration_validator.py [hunk 18]
```diff
@@ -120,3 +119,11 @@ def __verify_eval_loop_configuration(self, model, eval_loop_name):
...
                 f'you defined a {step_name} but have no {loader_name}. Skipping {eval_loop_name} loop'
             )
+
+    def check_invalid_loss_strategy(self):
+        values = [v.value for v in InvalidLossStrategy]
+        invalid_loss_strategy = self.trainer.get_model().invalid_loss_strategy
+        if invalid_loss_strategy not in values:
+            raise MisconfigurationException(
+                f'LightningModule `invalid_loss_strategy` property should be within {values}. Provided {invalid_loss_strategy}'
+            )
```
Comment (carmocca) [inline pytorch_lightning/trainer/configuration_validator.py:129]: Remove this and do it in the setter as mentioned above

```suggestion
```

File: pytorch_lightning/plugins/ddp_comm_hooks/allreduce_invalid.py [hunk 19]
```diff
@@ -0,0 +1,98 @@
...
+        return partial(init_hook,
+                       accumulate_grad_batches=trainer.accumulate_grad_batches,
+                       world_size=trainer.world_size)
+
+    def update_allreduce_hook_with_invalid_tensors(trainer, state):
+        state["should_accumulate"] = trainer.train_loop.should_accumulate()
+
+    def allreduce_hook_with_invalid_tensors(state: Dict, bucket: torch_distrib._GradBucket):
+        """
+        This DDP communication all_reduce hook enables to train with NaN or Inf losses
```
Comment (SeanNaren) [inline pytorch_lightning/plugins/ddp_comm_hooks/allreduce_invalid.py:47]:
```suggestion
        This DDP communication all_reduce hook enables to train with NaN or Inf losses by detecting invalid values and zeroing them for the optimizer step
```

File: pytorch_lightning/trainer/training_loop.py [hunk 20]
```diff
@@ -856,6 +860,38 @@ def backward(self, result, optimizer, opt_idx, *args, **kwargs):
             # track gradients
             self.track_and_norm_grad(optimizer=optimizer)
 
+    def should_return_on_invalid_result(self, result):
+        is_result_none = result is None
+        if isinstance(self.trainer.model, LightningDistributedDataParallel):
```
Comment (SeanNaren) [inline pytorch_lightning/trainer/training_loop.py:865]: Can we encompass this check in the accelerator somehow?

```if self.trainer.accelerator_backend.support_invalid_loss_strategy```?

File: pytorch_lightning/trainer/training_loop.py [hunk 21]
```diff
@@ -856,6 +860,38 @@ def backward(self, result, optimizer, opt_idx, *args, **kwargs):
...
+            skip_on_at_leat_one = model_ref.invalid_loss_strategy == InvalidLossStrategy.SKIP_IF_ANY
+            never_skip = model_ref.invalid_loss_strategy == InvalidLossStrategy.NEVER_SKIP
+
+            if never_skip and is_result_none:
+                raise MisconfigurationException(
+                    "invalid_loss_strategy ``never_skip`` doesn't support returning None for training_step. "
+                    "Hint: Either switch to ``skip_if_any`` or return a Nan or Inf loss directly"
+                )
+
+            if not skip_on_at_leat_one:
```
Comment (SeanNaren) [inline pytorch_lightning/trainer/training_loop.py:876]:
```suggestion
            if not skip_on_atleast_one:
```

File: pytorch_lightning/trainer/training_loop.py [hunk 22]
```diff
@@ -856,6 +860,38 @@ def backward(self, result, optimizer, opt_idx, *args, **kwargs):
             # track gradients
             self.track_and_norm_grad(optimizer=optimizer)
 
+    def should_return_on_invalid_result(self, result):
+        is_result_none = result is None
+        if isinstance(self.trainer.model, LightningDistributedDataParallel):
+            model_ref = self.trainer.get_model()
+            skip_on_at_leat_one = model_ref.invalid_loss_strategy == InvalidLossStrategy.SKIP_IF_ANY
```
Comment (SeanNaren) [inline pytorch_lightning/trainer/training_loop.py:867]:
```suggestion
            skip_on_atleast_one = model_ref.invalid_loss_strategy == InvalidLossStrategy.SKIP_IF_ANY
```

**Comment by Borda** [comment]
btw, the todo list is very long, shall we split the work on multiple PRs, easier to review and let it land...

**Comment by carmocca** [comment]
> btw, the todo list is very long, shall we split the work on multiple PRs, easier to review and let it land...

I agree, but still lots of work to do. We can keep developing here and chop it up into pieces when it is ready for general review

**Comment by edenlightning** [comment]
@tchaton is there anything left TODO here or can you close?

**Comment by carmocca** [comment]
> @tchaton is there anything left TODO here or can you close?

We need to address all the TODOs mentioned at the top. So many things left

**Comment by Borda** [comment]
seem this quite old PR with several handerts of commits behind the master, consider finishing it or closing as most likely the conflicts will make the PR challenging to finish... :rabbit:

**Comment by carmocca** [comment]
We can close this for now and use a new PR later.

**Comment by sailordiary** [comment]
Hi, is this issue being tracked elsewhere?

**Comment by carmocca** [comment]
There's the issue links at the top:

> Fixes #4956, #5243 (DDP), Fixes #4524 (AMP)
